File: tests/lsp/agreement_test.ts
import { mkdtempSync, writeFileSync } from "fs";
import { tmpdir } from "os";
import { join } from "path";
import { beforeAll, afterAll, expect, test } from "vitest";
import { buildGenctlBinary, runCli } from "../helpers/cli.ts";
import { at, Doc, Lsp, useWorkspace } from "./helpers.ts";

// The two binaries answer "what is this slot" from ONE place. `genctl schema type <process>`
// lists every slot with its summary, and a hover on a slot's key prints the same summary — both
// are `validation.TypeSlots` rendered by `schema.Summary`, reached through two front doors.
//
// This pins the front doors. The rule that made them disagree lived one level down, on a KEY
// inside a slot, and is pinned at that level by `TestKeyHoverIsTheCLIsOwnAnswer` in
// internal/lsp — in Go, because the CLI prints a document there and a summary is not derivable
// from one out here without re-implementing Summary.

let lsp: Lsp;
let dir: string;
beforeAll(async () => {
  useWorkspace();
  dir = mkdtempSync(join(tmpdir(), "genroc_agree_"));
  lsp = await Lsp.start();
}, 60_000);
afterAll(async () => lsp?.stop());

const OFFLINE = { GENROC_SERVER: "http://127.0.0.1:1" };

/** Writes the definition to disk (the CLI reads a file) and hands back both handles. */
function both(name: string, lines: string[]): { doc: Doc; file: string } {
  const file = join(dir, `${name}.genroc.yaml`);
  const text = [`name: ${name}`, ...lines].join("\n") + "\n";
  writeFileSync(file, text, "utf8");
  return { doc: { uri: `file://${file}`, text }, file };
}

/** The CLI's summary for one slot address, off the listing. */
function listed(name: string, file: string, address: string): string {
  const r = runCli(buildGenctlBinary(), ["schema", "type", name, "-f", file], OFFLINE);
  expect(r.stderr).toBe("");
  const line = r.stdout.split("\n").find((l) => l.startsWith(address + " "));
  expect(line, `no listing line for ${address} in:\n${r.stdout}`).toBeDefined();
  return line!.slice(address.length).trim();
}

/** The type inside a hover's code span. */
function hovered(h: string): string {
  const m = /`([^`]+)`/.exec(h);
  expect(m, `no type in hover: ${h}`).not.toBeNull();
  return m![1];
}

const cases: { name: string; lines: string[]; address: string; fragment: string }[] = [
  {
    // The pair that disagreed one level down: nullable expression, non-nullable declaration.
    name: "agree_repair",
    lines: [
      "input_schema:",
      "  type: object",
      '  properties: { n: { type: [number, "null"] } }',
      "tasks:",
      "  - id: t",
      "    action:",
      "      type: child",
      "      name: nope",
      "      input:",
      "        v: '$: input.n'",
      "      input_schema: { type: object, properties: { v: { type: number } } }",
      "    switch: [{ goto: end }]",
      "output: { ok: true }",
    ],
    address: "tasks.t.action.input",
    fragment: "      <^input>:",
  },
  {
    // A generic child's payload, declared as the top type.
    name: "agree_unknown",
    lines: [
      "tasks:",
      "  - id: t",
      "    action:",
      "      type: child",
      "      name: nope",
      "      input:",
      "        payload:",
      "          who: 'x'",
      "      input_schema: { type: object, properties: { payload: { description: opaque } } }",
      "    switch: [{ goto: end }]",
      "output: { ok: true }",
    ],
    address: "tasks.t.action.input",
    fragment: "      <^input>:",
  },
  {
    // One of the two slots that was not in the type view at all, so the CLI could not answer.
    name: "agree_query",
    lines: [
      "tasks:",
      "  - id: t",
      "    action:",
      "      type: fetch",
      "      url: http://x.invalid/y",
      "      method: get",
      "      query:",
      "        page: 2",
      "      query_schema: { type: object, properties: { page: { type: number } } }",
      "    switch: [{ goto: end }]",
      "output: { ok: true }",
    ],
    address: "tasks.t.action.query",
    fragment: "      <^query>:",
  },
  {
    // A slot the definition HANDS BACK, where the declaration is what both publish.
    name: "agree_output",
    lines: [
      "tasks:",
      "  - id: t",
      "    output: { v: 1, w: 2 }",
      "    output_schema: { type: object, properties: { v: { type: number }, w: { type: number } }, required: [v] }",
      "    switch: [{ goto: end }]",
      "output: { ok: true }",
    ],
    address: "tasks.t.output",
    fragment: "    <^output>: { v: 1, w: 2 }",
  },
];

for (const c of cases) {
  test(`${c.name}: the CLI's listing and the hover print one summary`, async () => {
    const { doc, file } = both(c.name, c.lines);
    const cli = listed(c.name, file, c.address);
    const hover = hovered(await lsp.hover(at(c.fragment, doc)));
    expect(hover, `hover ${JSON.stringify(hover)} vs CLI ${JSON.stringify(cli)}`).toBe(cli);
  });
}

File: internal/lsp/agreement_test.go
package lsp

import (
	"strings"
	"testing"

	"genroc/internal/validation"
)

// The one claim the type unification makes, at the level where it was false: a KEY inside a
// shape hovers as the type validation computed for its slot, so the editor's answer is the
// CLI's own. For a day it was not — `genctl schema type` published the inferred side while
// hover applied a rule of its own, and one key read `number|null` in the CLI and `number` in
// the editor. tests/lsp/agreement_test.ts pins the two binaries at the SLOT; this is the key,
// in Go because the CLI's lookup is a function call here and a printed document out there.
//
//	 1 name: agree
//	 2 input_schema:
//	 3   type: object
//	 4   properties: { n: { type: [number, "null"] } }
//	 5 tasks:
//	 6   - id: t
//	 7     action:
//	 8       type: child
//	 9       name: nope
//	10       input:
//	11         v: '$: input.n'
//	12       input_schema: { type: object, properties: { v: { type: number, description: prose } } }
//	13     switch: [{ goto: end }]
//	14 output: { ok: true }
const agreeDoc = `name: agree
input_schema:
  type: object
  properties: { n: { type: [number, "null"] } }
tasks:
  - id: t
    action:
      type: child
      name: nope
      input:
        v: '$: input.n'
      input_schema: { type: object, properties: { v: { type: number, description: prose } } }
    switch: [{ goto: end }]
output: { ok: true }
`

func TestKeyHoverIsTheCLIsOwnAnswer(t *testing.T) {
	docs, err := parseDocuments(agreeDoc, "")
	if err != nil || len(docs) != 1 {
		t.Fatalf("parse: %v", err)
	}
	def, ok := docs[0].definition()
	if !ok {
		t.Fatal("the document did not decode")
	}
	slots, err := validation.TypeSlots(def)
	if err != nil {
		t.Fatalf("TypeSlots: %v", err)
	}
	// What `genctl schema type agree tasks.t.action.input` prints, and then the member read off
	// it — the two steps shapeKeyHover takes, so a rule added on EITHER side breaks this. Not the
	// key's own address: `SlotAt` walks it the way an expression would, so an optional member
	// comes back nullable there, which is the read view and not a description of the key.
	parent, found, err := validation.SlotAt(slots, "tasks.t.action.input")
	if err != nil || !found {
		t.Fatalf("the slot is not in the type view: %v", err)
	}
	parent = unwrap(parent)
	prop, declared := parent.Properties()["v"]
	if !declared {
		t.Fatal("the slot's type carries no `v`")
	}

	// Line 11 is `        v: '$: input.n'`; column 9 is inside `v`.
	md := hoverOf(t, agreeDoc, 11, 9)
	if want := "`" + prop.Summary() + "`"; !strings.Contains(md, want) {
		t.Errorf("hover %q does not carry the CLI's type %s", md, want)
	}
	if absent := parent.MayBeAbsent("v"); absent != strings.Contains(md, "**v?**") {
		t.Errorf("hover %q and the type view disagree about whether `v` may be absent (%v)", md, absent)
	}
	if !strings.Contains(md, "prose") {
		t.Errorf("hover %q lost the declaration's prose, which the conformed type carries", md)
	}

	// The premise, so this cannot pass by both sides being wrong together. The conform removes
	// the null and the key with it: `number`, may be absent — not the `number|null` the
	// expression types, which is what the CLI answered before the type was computed once.
	if s := prop.Summary(); s != "number" {
		t.Fatalf("premise gone: the slot's `v` reads %q; the repair this pins is the null removal", s)
	}
	if !parent.MayBeAbsent("v") {
		t.Fatal("premise gone: a removed null must leave the key may-be-absent")
	}
}
